test: conformance testing with Prism (#79)
govuk-one-login/mobile-wallet-status-list-mock: test/conformance/README.md

# Conformance Tests

Prism proxy-mode conformance tests that verify the deployed service matches its OAS

> contract end-to-end.

Reviewer: ```suggestion
Prism proxy mode conformance tests verify that the deployed Status List Mock service matches its OpenAPI specification (OAS) contract end-to-end.
```

## How it works

Prism runs in proxy mode between the test and the real service. It validates every
request and response against the OAS spec and rejects anything that violates the
contract before it reaches (or leaves) the upstream:

```
Test (Jest) → Prism proxy (OAS enforcer) → deployed service → Prism → Test
```

The test code only asserts HTTP status codes and response shapes. Prism does the OAS
enforcement — no spec logic is duplicated in the tests.

## Structure

```
test/conformance/
  helpers/                      Wait for port utility
  testSuits/                    Factory functions, add test cases here
  runners/                      Config and runtime setup 
    local/                      Runs against SAM local (localhost:3000)

>     ci/                         Runs against deployed service (UPSTREAM_URL)

Reviewer: ```suggestion
    ci/                         Runs against deployed service (UPSTREAM_URL)
    local/                      Runs against locally running service (localhost:3000)
```

```

Each suite is a factory function. Runner files call the factory with environment-specific
setup (upstream URL, health check, warm-up). This keeps the test cases defined once and
shared between local and CI runners.

## Test suites

| Suite | Spec | Port | Covers |
|---|---|---|---|
| `issueConformanceSuite.ts` | `crs-private-spec.yaml` | 4010 | `POST /issue` — Content-Type enforcement, 200 response schema |
| `revokeConformanceSuite.ts` | `crs-private-spec.yaml` | 4010 | `POST /revoke` — Content-Type enforcement, 202 response schema |

> | `publicConformanceSuite.ts` | `crs-public-spec.yaml` | 4011 | `GET /t/{id}` — response Content-Type, schema validation |

Reviewer: If you want the order in which these files are mentioned to match the repository structure, `publicConformanceSuite.ts` needs to come before `revokeConformanceSuite.ts`.

Author: The objective is to show which suit uses which OAS and port. So it makes sense to keep as is.

the spec copy first.** The conformance failure is likely a symptom of the drift.

### How Prism tests and OAS Diff work together

Drift detection ensures the spec copies in this repo stay in sync with the upstream CRS
spec. Conformance tests ensure the deployed service honours those specs. If the real CRS
contract changes, drift detection catches it. If a code or infrastructure change breaks
the mock's behaviour, conformance tests catch it.

## When the conformance tests fail

A Slack alert is sent to the OP channel on failure. The cause will be one of:

- **422 from Prism** — request violated the OAS spec (e.g. wrong Content-Type). Check the test or spec.
- **Unexpected status code** — service returned an error (e.g. 500 instead of 200). Check the service or infrastructure.
- **Response schema mismatch** — service response does not match the OAS schema. Compare the response against `crs-private-spec.yaml` or `crs-public-spec.yaml`.
- **Prism failed to start** — upstream unreachable. Verify the deployment completed and `UPSTREAM_URL` is correct.

If `check-oas-for-drift` has also failed, update the spec copy first — the two failures are likely related.

> 

Reviewer: ```suggestion
```

## Known gaps

> Error response cases (400, 401, 403, 404 for `/issue` and `/revoke`) are not yet covered.

Reviewer: ```suggestion
Error response cases (400, 401, 403, 404 for `/issue` and `/revoke`) are not covered because the mock only implements the happy path. Testing error scenarios would require the mock handlers to implement JWT claim validation and return the appropriate error responses.
```
I'd say this instead because I'm not aware of upcoming work to change the mock to cover unhappy path scenarios as well.
